# 시각화!

더 뽑아먹을 거리가 있는지 찾아보기 위해 오늘도 데이터를 살펴봅니다.


In [115]:
import ast
import itertools
from operator import itemgetter
from pprint import pprint

import numpy as np
import pandas as pd

In [116]:
chars = pd.read_excel('./datasets/characters.xlsx')
chars.info()
chars.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 452 entries, 0 to 451
Data columns (total 24 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   451 non-null    object 
 1   name_ko              452 non-null    object 
 2   name_en              452 non-null    object 
 3   special              452 non-null    bool   
 4   gender               452 non-null    object 
 5   gender_asia          452 non-null    object 
 6   species              391 non-null    object 
 7   birthday             452 non-null    object 
 8   birthday_month       452 non-null    int64  
 9   birthday_day         452 non-null    int64  
 10  birthday_yday        452 non-null    int64  
 11  hobby                391 non-null    object 
 12  personality          391 non-null    object 
 13  personality_subtype  391 non-null    object 
 14  colors               391 non-null    object 
 15  styles               391 non-null    obj

id name_ko  name_en  special  gender gender_asia species birthday  \
0  admiral      일섭  Admiral    False    Male        Male       새    01-27   
1   agents      2호  Agent S    False  Female      Female     다람쥐    07-02   
2    agnes    아그네스    Agnes    False  Female      Female      돼지    04-21   
3       al      우락       Al    False    Male        Male     고릴라    10-18   
4  alfonso    알베르트  Alfonso    False    Male        Male      악어    06-09   

   birthday_month  birthday_day  ...          colors          styles tier  \
0               1            27  ...  ['검정색', '파랑색']           ['쿨']  6.0   
1               7             2  ...  ['검정색', '파랑색']   ['심플', '액티브']  6.0   
2               4            21  ...  ['하양색', '핑크색']  ['심플', '엘레강스']  4.0   
3              10            18  ...  ['하양색', '빨강색']         ['액티브']  6.0   
4               6             9  ...  ['빨강색', '파랑색']          ['심플']  6.0   

   rank catchphrase                                favorite_saying  \
0  89.0     aye aye                         Only quitters give up.   
1  45.0    sidekick          You gotta put the pedal to the metal!   
2  26.0     snuffle                         You reap what you sow.   
3  19.0     ayyyeee  Life is a buffet. Always go back for seconds.   
4   6.0    it'sa me                 Slow and steady wins the race.   

   favorite_song                                        image_house  \
0     Steep Hill  https://acnhcdn.com/drivesync/render/houses/br...   
1  Go K.K. Rider  https://acnhcdn.com/drivesync/render/houses/sq...   
2     K.K. House  https://acnhcdn.com/drivesync/render/houses/pi...   
3  Go K.K. Rider  https://acnhcdn.com/drivesync/render/houses/go...   
4    Forest Life  https://acnhcdn.com/drivesync/render/houses/cr...   

                                     image_icon  \
0  https://acnhcdn.com/latest/NpcIcon/brd06.png   
1  https://acnhcdn.com/latest/NpcIcon/squ05.png   
2  https://acnhcdn.com/latest/NpcIcon/pig17.png   
3  https://acnhcdn.com/latest/NpcIcon/gor08.png   
4  https://acnhcdn.com/latest/NpcIcon/crd00.png   

                                         image_photo  
0  https://acnhcdn.com/latest/NpcBromide/NpcNmlBr...  
1  https://acnhcdn.com/latest/NpcBromide/NpcNmlSq...  
2  https://acnhcdn.com/latest/NpcBromide/NpcNmlPi...  
3  https://acnhcdn.com/latest/NpcBromide/NpcNmlGo...  
4  https://acnhcdn.com/latest/NpcBromide/NpcNmlCr...  

[5 rows x 24 columns]

## 뭘 할 수 있는지 보기

생각하기 귀찮으니까 컴퓨터한테 시킵니다.


In [117]:
# list(itertools.permutations(['성격', '취미', '색깔', '성별', '스타일'], 2))
list(itertools.permutations(['성격', '취미', '성별',], 2))

[('성격', '취미'),
 ('성격', '성별'),
 ('취미', '성격'),
 ('취미', '성별'),
 ('성별', '성격'),
 ('성별', '취미')]

## 성격별 xx

성격별로 뭔가 달라지는 게 있을까요


### 성격별 취미

```
personality  hobby
느끼함          교육       11
             운동        4
             음악       14
             자연        5
단순 활발        놀이        7
             운동        6
             음악       11
먹보           교육        1
             놀이       32
             운동        4
             자연       23
무뚝뚝          교육       19
             놀이        2
             운동       10
             음악       13
             자연       11
성숙함          교육       10
             운동        3
             음악        8
             자연        2
             패션       32
아이돌          놀이        9
             운동        3
             음악        5
             자연        1
             패션       31
운동광          교육        1
             놀이       15
             운동       36
             음악        3
친절함          교육       22
             음악       10
             자연       24
             패션        3
dtype: int64
```

In [118]:
chars.groupby(['personality', 'hobby']).size()

personality  hobby
느끼함          교육       11
             운동        4
             음악       14
             자연        5
단순 활발        놀이        7
             운동        6
             음악       11
먹보           교육        1
             놀이       32
             운동        4
             자연       23
무뚝뚝          교육       19
             놀이        2
             운동       10
             음악       13
             자연       11
성숙함          교육       10
             운동        3
             음악        8
             자연        2
             패션       32
아이돌          놀이        9
             운동        3
             음악        5
             자연        1
             패션       31
운동광          교육        1
             놀이       15
             운동       36
             음악        3
친절함          교육       22
             음악       10
             자연       24
             패션        3
dtype: int64

### 성격별 성별

```
personality  gender_asia
느끼함          Male           34
단순 활발        Female         24
먹보           Male           60
무뚝뚝          Male           55
성숙함          Female         55
아이돌          Female         49
운동광          Male           55
친절함          Female         59
dtype: int64
```

In [119]:
chars.groupby(['personality', 'gender_asia']).size()

personality  gender_asia
느끼함          Male           34
단순 활발        Female         24
먹보           Male           60
무뚝뚝          Male           55
성숙함          Female         55
아이돌          Female         49
운동광          Male           55
친절함          Female         59
dtype: int64

### 취미별 성격

```
hobby  personality
교육     느끼함            11
       먹보              1
       무뚝뚝            19
       성숙함            10
       운동광             1
       친절함            22
놀이     단순 활발           7
       먹보             32
       무뚝뚝             2
       아이돌             9
       운동광            15
운동     느끼함             4
       단순 활발           6
       먹보              4
       무뚝뚝            10
       성숙함             3
       아이돌             3
       운동광            36
음악     느끼함            14
       단순 활발          11
       무뚝뚝            13
       성숙함             8
       아이돌             5
       운동광             3
       친절함            10
자연     느끼함             5
       먹보             23
       무뚝뚝            11
       성숙함             2
       아이돌             1
       친절함            24
패션     성숙함            32
       아이돌            31
       친절함             3
dtype: int64
```


In [120]:
chars.groupby(['hobby', 'personality']).size()


hobby  personality
교육     느끼함            11
       먹보              1
       무뚝뚝            19
       성숙함            10
       운동광             1
       친절함            22
놀이     단순 활발           7
       먹보             32
       무뚝뚝             2
       아이돌             9
       운동광            15
운동     느끼함             4
       단순 활발           6
       먹보              4
       무뚝뚝            10
       성숙함             3
       아이돌             3
       운동광            36
음악     느끼함            14
       단순 활발          11
       무뚝뚝            13
       성숙함             8
       아이돌             5
       운동광             3
       친절함            10
자연     느끼함             5
       먹보             23
       무뚝뚝            11
       성숙함             2
       아이돌             1
       친절함            24
패션     성숙함            32
       아이돌            31
       친절함             3
dtype: int64

### 취미별 성별

```
hobby  gender_asia
교육     Female         32
       Male           32
놀이     Female         16
       Male           49
운동     Female         12
       Male           54
음악     Female         34
       Male           30
자연     Female         27
       Male           39
패션     Female         66
dtype: int64
```

In [121]:
chars.groupby(['hobby', 'gender_asia']).size()

hobby  gender_asia
교육     Female         32
       Male           32
놀이     Female         16
       Male           49
운동     Female         12
       Male           54
음악     Female         34
       Male           30
자연     Female         27
       Male           39
패션     Female         66
dtype: int64

### 성별별 취미

```
gender_asia  hobby
Female       교육       32
             놀이       16
             운동       12
             음악       34
             자연       27
             패션       66
Male         교육       32
             놀이       49
             운동       54
             음악       30
             자연       39
dtype: int64
```

In [122]:
chars.groupby(['gender_asia', 'hobby']).size()

gender_asia  hobby
Female       교육       32
             놀이       16
             운동       12
             음악       34
             자연       27
             패션       66
Male         교육       32
             놀이       49
             운동       54
             음악       30
             자연       39
dtype: int64

### 성별별 성격

이게 의미가 있을 수도 있는 이유는 성별별로 성격이 따로있다는 거는 게임을 안해보면 절대 상상도 못할 일이기 때문입니다.

```
gender_asia  personality
Female       단순 활발          24
             성숙함            55
             아이돌            49
             친절함            59
Male         느끼함            34
             먹보             60
             무뚝뚝            55
             운동광            55
dtype: int64
```

In [123]:
chars.groupby(['gender_asia', 'personality']).size()

gender_asia  personality
Female       단순 활발          24
             성숙함            55
             아이돌            49
             친절함            59
Male         느끼함            34
             먹보             60
             무뚝뚝            55
             운동광            55
dtype: int64

### 성별별 취미

```
gender_asia  hobby
Female       교육       32
             놀이       16
             운동       12
             음악       34
             자연       27
             패션       66
Male         교육       32
             놀이       49
             운동       54
             음악       30
             자연       39
dtype: int64
```

In [124]:
chars.groupby(['gender_asia', 'hobby']).size()

gender_asia  hobby
Female       교육       32
             놀이       16
             운동       12
             음악       34
             자연       27
             패션       66
Male         교육       32
             놀이       49
             운동       54
             음악       30
             자연       39
dtype: int64

## 인기도?

tier와 rank 2가지가 있는데 rank까지는 쓸 필요가 없었으면 합니다.
너무 머리 아프니까...


### 성별 분포

깜박할뻔 했는데 성별의 분포는 해석에 꽤 중요합니다.


In [125]:
chars.groupby('gender_asia').size()

gender_asia
Female    205
Male      247
dtype: int64

남 247 : 여 205

근데 지금 스페셜 주민도 끼어 있습니다.

얘네는 생일 말곤 속성이 없어서 도움이 안됩니다.


In [126]:
chars_v = pd.read_excel('./datasets/characters-similarity.xlsx')
chars_v['colors'] = chars_v['colors'].apply(ast.literal_eval)
chars_v['styles'] = chars_v['styles'].apply(ast.literal_eval)
chars_v.head()

id  name_en name_ko                                    image_icon  \
0  admiral  Admiral      일섭  https://acnhcdn.com/latest/NpcIcon/brd06.png   
1   agents  Agent S      2호  https://acnhcdn.com/latest/NpcIcon/squ05.png   
2    agnes    Agnes    아그네스  https://acnhcdn.com/latest/NpcIcon/pig17.png   
3       al       Al      우락  https://acnhcdn.com/latest/NpcIcon/gor08.png   
4  alfonso  Alfonso    알베르트  https://acnhcdn.com/latest/NpcIcon/crd00.png   

     species  gender birthday  birthday_month  birthday_day personality hobby  \
0       Bird    Male    01-27               1            27         무뚝뚝    자연   
1   Squirrel  Female    07-02               7             2         아이돌    운동   
2        Pig  Female    04-21               4            21       단순 활발    놀이   
3    Gorilla    Male    10-18              10            18          먹보    운동   
4  Alligator    Male    06-09               6             9          먹보    놀이   

       styles      colors  
0         [쿨]  [검정색, 파랑색]  
1   [심플, 액티브]  [검정색, 파랑색]  
2  [심플, 엘레강스]  [하양색, 핑크색]  
3       [액티브]  [하양색, 빨강색]  
4        [심플]  [빨강색, 파랑색]

In [127]:
chars_v.groupby('gender').size()

gender
Female    187
Male      204
dtype: int64

f: 187 vs. m: 204

대체로 비슷비슷한 것 같습니다.


### 성별/인기도

여자가 남자보다 약간 적은 것을 감안하면 여자일수록 약간 인기가 많은 경향이 있습니다.


In [128]:
chars_v = chars[chars['special'].eq(False)]

In [129]:
chars_v.groupby(['gender', 'tier']).size()

gender  tier
Female  1.0      6
        2.0     14
        3.0     18
        4.0     30
        5.0     58
        6.0     61
Male    1.0      6
        2.0     10
        3.0     11
        4.0     25
        5.0     55
        6.0     96
dtype: int64

In [130]:
chars_v.groupby(['tier', 'gender']).size()

tier  gender
1.0   Female     6
      Male       6
2.0   Female    14
      Male      10
3.0   Female    18
      Male      11
4.0   Female    30
      Male      25
5.0   Female    58
      Male      55
6.0   Female    61
      Male      96
dtype: int64

### 성격 / 인기도

성격과 인기도는 뚜렷한 상관 관계는 없는 것 같습니다.


```
personality  tier
느끼함          1.0      2
             2.0      2
             3.0      2
             4.0      6
             5.0      5
             6.0     17
단순 활발        2.0      1
             3.0      5
             4.0      6
             5.0      8
             6.0      4
먹보           1.0      4
             2.0      4
             3.0      2
             4.0      9
             5.0     19
             6.0     21
무뚝뚝          2.0      2
             3.0      4
             4.0      5
             5.0     13
             6.0     31
성숙함          1.0      2
             2.0      1
             3.0      1
             4.0      6
             5.0     15
             6.0     30
아이돌          2.0      6
             3.0      4
             4.0     10
             5.0     16
             6.0     13
운동광          2.0      2
             3.0      3
             4.0      5
             5.0     18
             6.0     27
친절함          1.0      4
             2.0      6
             3.0      8
             4.0      8
             5.0     19
             6.0     14
dtype: int64
```

```
tier  personality
1.0   느끼함             2
      먹보              4
      성숙함             2
      친절함             4
2.0   느끼함             2
      단순 활발           1
      먹보              4
      무뚝뚝             2
      성숙함             1
      아이돌             6
      운동광             2
      친절함             6
3.0   느끼함             2
      단순 활발           5
      먹보              2
      무뚝뚝             4
      성숙함             1
      아이돌             4
      운동광             3
      친절함             8
4.0   느끼함             6
      단순 활발           6
      먹보              9
      무뚝뚝             5
      성숙함             6
      아이돌            10
      운동광             5
      친절함             8
5.0   느끼함             5
      단순 활발           8
      먹보             19
      무뚝뚝            13
      성숙함            15
      아이돌            16
      운동광            18
      친절함            19
6.0   느끼함            17
      단순 활발           4
      먹보             21
      무뚝뚝            31
      성숙함            30
      아이돌            13
      운동광            27
      친절함            14
dtype: int64
```

In [131]:
chars_v.groupby(['personality', 'tier']).size()

personality  tier
느끼함          1.0      2
             2.0      2
             3.0      2
             4.0      6
             5.0      5
             6.0     17
단순 활발        2.0      1
             3.0      5
             4.0      6
             5.0      8
             6.0      4
먹보           1.0      4
             2.0      4
             3.0      2
             4.0      9
             5.0     19
             6.0     21
무뚝뚝          2.0      2
             3.0      4
             4.0      5
             5.0     13
             6.0     31
성숙함          1.0      2
             2.0      1
             3.0      1
             4.0      6
             5.0     15
             6.0     30
아이돌          2.0      6
             3.0      4
             4.0     10
             5.0     16
             6.0     13
운동광          2.0      2
             3.0      3
             4.0      5
             5.0     18
             6.0     27
친절함          1.0      4
             2.0      6
             3.0      

In [132]:
chars_v.groupby(['tier', 'personality']).size()

tier  personality
1.0   느끼함             2
      먹보              4
      성숙함             2
      친절함             4
2.0   느끼함             2
      단순 활발           1
      먹보              4
      무뚝뚝             2
      성숙함             1
      아이돌             6
      운동광             2
      친절함             6
3.0   느끼함             2
      단순 활발           5
      먹보              2
      무뚝뚝             4
      성숙함             1
      아이돌             4
      운동광             3
      친절함             8
4.0   느끼함             6
      단순 활발           6
      먹보              9
      무뚝뚝             5
      성숙함             6
      아이돌            10
      운동광             5
      친절함             8
5.0   느끼함             5
      단순 활발           8
      먹보             19
      무뚝뚝            13
      성숙함            15
      아이돌            16
      운동광            18
      친절함            19
6.0   느끼함            17
      단순 활발           4
      먹보             21
      무뚝뚝            31
      성숙함            3

### 취미 / 인기도

취미별 인기도도 뚜렷한 경향은 없습니다.
얼핏 보면 자연이 꽤 많아 보이기도 하지만 확신할 수준은 아닙니다.

```
hobby  tier
교육     1.0      2
       2.0      5
       3.0      4
       4.0     10
       5.0     13
       6.0     30
놀이     1.0      1
       2.0      6
       3.0      6
       4.0     12
       5.0     15
       6.0     24
운동     2.0      2
       3.0      2
       4.0      4
       5.0     20
       6.0     38
음악     1.0      3
       2.0      7
       3.0      6
       4.0      9
       5.0     19
       6.0     20
자연     1.0      6
       2.0      3
       3.0      8
       4.0     10
       5.0     19
       6.0     20
패션     2.0      1
       3.0      3
       4.0     10
       5.0     27
       6.0     25
dtype: int64
```

```
tier  hobby
1.0   교육        2
      놀이        1
      음악        3
      자연        6
2.0   교육        5
      놀이        6
      운동        2
      음악        7
      자연        3
      패션        1
3.0   교육        4
      놀이        6
      운동        2
      음악        6
      자연        8
      패션        3
4.0   교육       10
      놀이       12
      운동        4
      음악        9
      자연       10
      패션       10
5.0   교육       13
      놀이       15
      운동       20
      음악       19
      자연       19
      패션       27
6.0   교육       30
      놀이       24
      운동       38
      음악       20
      자연       20
      패션       25
dtype: int64
```

In [133]:
chars_v.groupby(['hobby', 'tier']).size()

hobby  tier
교육     1.0      2
       2.0      5
       3.0      4
       4.0     10
       5.0     13
       6.0     30
놀이     1.0      1
       2.0      6
       3.0      6
       4.0     12
       5.0     15
       6.0     24
운동     2.0      2
       3.0      2
       4.0      4
       5.0     20
       6.0     38
음악     1.0      3
       2.0      7
       3.0      6
       4.0      9
       5.0     19
       6.0     20
자연     1.0      6
       2.0      3
       3.0      8
       4.0     10
       5.0     19
       6.0     20
패션     2.0      1
       3.0      3
       4.0     10
       5.0     27
       6.0     25
dtype: int64

In [134]:
chars_v.groupby(['tier', 'hobby']).size()

tier  hobby
1.0   교육        2
      놀이        1
      음악        3
      자연        6
2.0   교육        5
      놀이        6
      운동        2
      음악        7
      자연        3
      패션        1
3.0   교육        4
      놀이        6
      운동        2
      음악        6
      자연        8
      패션        3
4.0   교육       10
      놀이       12
      운동        4
      음악        9
      자연       10
      패션       10
5.0   교육       13
      놀이       15
      운동       20
      음악       19
      자연       19
      패션       27
6.0   교육       30
      놀이       24
      운동       38
      음악       20
      자연       20
      패션       25
dtype: int64

### 특수 주민은??

특수 주민은 다른 정보는 없어도 최소한 성별은 볼 수 있습니다.
성별을 검사해 봅시다.


In [135]:
chars_s = chars[chars['special'].eq(True)]

In [136]:
chars_s.groupby(['gender', 'tier']).size()

Series([], dtype: int64)

In [137]:
chars_s.groupby(['tier', 'gender']).size()

Series([], dtype: int64)

이런.. 특수 npc는 인기도도 없습니다.

인기는 많을텐데...


### 인기도의 종별 분포?

종은 외모에 상당히 결정적입니다.


--> 뭔가 경향이 있긴 한데 종이 너무 많아서 도저히 알 수가 없다.


In [138]:
chars_v.groupby(['species', 'tier']).size()
print(_.to_string())

Series([], )


In [139]:
chars_v.groupby(['tier', 'species']).size()
print(_.to_string())

Series([], )


종별로 세보는 게 답인 것 같습니다.


In [140]:
num_cats = len(chars_v[chars_v['species'] == '고양이'])
chars_v[chars_v['species'] == '고양이'].groupby('tier').size() * 100 / num_cats

tier
1.0     8.695652
2.0    26.086957
3.0    17.391304
4.0    17.391304
5.0    21.739130
6.0     8.695652
dtype: float64

In [141]:
species = list(chars_v['species'].unique())

In [142]:
species.sort()

In [143]:
df = pd.DataFrame(index=species, columns=range(1, 7))
df.head()

1    2    3    4    5    6
개     NaN  NaN  NaN  NaN  NaN  NaN
개구리   NaN  NaN  NaN  NaN  NaN  NaN
개미핥기  NaN  NaN  NaN  NaN  NaN  NaN
고릴라   NaN  NaN  NaN  NaN  NaN  NaN
고양이   NaN  NaN  NaN  NaN  NaN  NaN

In [144]:
chars_t = chars_v[chars_v['tier'].notnull()].copy()
chars_t['tier'] = chars_t['tier'].astype(int)


In [145]:
for index, row in df.iterrows():
  df.loc[index] = chars_t[chars_t['species'] == index].groupby('tier').size()


In [146]:
df = df.fillna(0).astype(int)

In [147]:
count = pd.DataFrame(chars_t.groupby('species').size(), columns=['count'])
df = df.merge(count, left_index=True, right_index=True)

In [148]:
df

1  2  3  4   5   6  count
개     0  2  3  4   4   3     16
개구리   0  1  1  1   8   7     18
개미핥기  0  0  0  0   3   4      7
고릴라   0  0  0  0   1   8      9
고양이   2  6  4  4   5   2     23
곰     0  0  0  3   3   9     15
꼬마곰   3  1  1  2   5   4     16
늑대    0  2  4  3   2   0     11
다람쥐   1  1  2  4   6   4     18
닭     0  0  0  0   4   5      9
독수리   0  1  0  0   3   5      9
돼지    0  0  0  1   1  12     14
말     0  1  1  1   4   8     15
문어    2  0  1  0   0   0      3
사슴    1  3  3  3   0   0     10
사자    0  0  0  1   1   5      7
새     0  0  0  0   3  10     13
수소    0  0  0  0   2   4      6
악어    0  0  0  2   1   4      7
암소    0  0  0  0   3   1      4
양     0  2  0  1   5   5     13
염소    1  0  2  2   0   3      8
오리    0  2  0  3   3   9     17
원숭이   0  0  0  1   4   3      8
쥐     0  0  0  1   4  10     15
캥거루   0  0  0  0   3   5      8
코끼리   0  0  1  2   3   5     11
코뿔소   1  0  0  1   3   1      6
코알라   0  0  1  3   5   0      9
타조    0  1  0  2   4   3     10
토끼    1  0  2  3  10   4     20
펭귄    0  1  0  4   4   4     13
하마    0  0  0  0   2   5      7
햄스터   0  0  3  1   2   2      8
호랑이   0  0  0  2   2   3      7

In [149]:
df_pct = df.copy()
df_pct.loc[:, 1:6] = df_pct.loc[:, 1:6].div(df_pct['count'], axis=0) * 100

In [150]:
df_pct

1          2          3          4          5          6  count
개      0.000000  12.500000  18.750000  25.000000  25.000000  18.750000     16
개구리    0.000000   5.555556   5.555556   5.555556  44.444444  38.888889     18
개미핥기   0.000000   0.000000   0.000000   0.000000  42.857143  57.142857      7
고릴라    0.000000   0.000000   0.000000   0.000000  11.111111  88.888889      9
고양이    8.695652  26.086957  17.391304  17.391304  21.739130   8.695652     23
곰      0.000000   0.000000   0.000000  20.000000  20.000000  60.000000     15
꼬마곰   18.750000   6.250000   6.250000  12.500000  31.250000  25.000000     16
늑대     0.000000  18.181818  36.363636  27.272727  18.181818   0.000000     11
다람쥐    5.555556   5.555556  11.111111  22.222222  33.333333  22.222222     18
닭      0.000000   0.000000   0.000000   0.000000  44.444444  55.555556      9
독수리    0.000000  11.111111   0.000000   0.000000  33.333333  55.555556      9
돼지     0.000000   0.000000   0.000000   7.142857   7.142857  85.714286     14
말      0.000000   6.666667   6.666667   6.666667  26.666667  53.333333     15
문어    66.666667   0.000000  33.333333   0.000000   0.000000   0.000000      3
사슴    10.000000  30.000000  30.000000  30.000000   0.000000   0.000000     10
사자     0.000000   0.000000   0.000000  14.285714  14.285714  71.428571      7
새      0.000000   0.000000   0.000000   0.000000  23.076923  76.923077     13
수소     0.000000   0.000000   0.000000   0.000000  33.333333  66.666667      6
악어     0.000000   0.000000   0.000000  28.571429  14.285714  57.142857      7
암소     0.000000   0.000000   0.000000   0.000000  75.000000  25.000000      4
양      0.000000  15.384615   0.000000   7.692308  38.461538  38.461538     13
염소    12.500000   0.000000  25.000000  25.000000   0.000000  37.500000      8
오리     0.000000  11.764706   0.000000  17.647059  17.647059  52.941176     17
원숭이    0.000000   0.000000   0.000000  12.500000  50.000000  37.500000      8
쥐      0.000000   0.000000   0.000000   6.666667  26.666667  66.666667     15
캥거루    0.000000   0.000000   0.000000   0.000000  37.500000  62.500000      8
코끼리    0.000000   0.000000   9.090909  18.181818  27.272727  45.454545     11
코뿔소   16.666667   0.000000   0.000000  16.666667  50.000000  16.666667      6
코알라    0.000000   0.000000  11.111111  33.333333  55.555556   0.000000      9
타조     0.000000  10.000000   0.000000  20.000000  40.000000  30.000000     10
토끼     5.000000   0.000000  10.000000  15.000000  50.000000  20.000000     20
펭귄     0.000000   7.692308   0.000000  30.769231  30.769231  30.769231     13
하마     0.000000   0.000000   0.000000   0.000000  28.571429  71.428571      7
햄스터    0.000000   0.000000  37.500000  12.500000  25.000000  25.000000      8
호랑이    0.000000   0.000000   0.000000  28.571429  28.571429  42.857143      7

In [151]:
df_pct.sort_values(by=[1, 2, 3, 4, 5], ascending=False)

1          2          3          4          5          6  count
문어    66.666667   0.000000  33.333333   0.000000   0.000000   0.000000      3
꼬마곰   18.750000   6.250000   6.250000  12.500000  31.250000  25.000000     16
코뿔소   16.666667   0.000000   0.000000  16.666667  50.000000  16.666667      6
염소    12.500000   0.000000  25.000000  25.000000   0.000000  37.500000      8
사슴    10.000000  30.000000  30.000000  30.000000   0.000000   0.000000     10
고양이    8.695652  26.086957  17.391304  17.391304  21.739130   8.695652     23
다람쥐    5.555556   5.555556  11.111111  22.222222  33.333333  22.222222     18
토끼     5.000000   0.000000  10.000000  15.000000  50.000000  20.000000     20
늑대     0.000000  18.181818  36.363636  27.272727  18.181818   0.000000     11
양      0.000000  15.384615   0.000000   7.692308  38.461538  38.461538     13
개      0.000000  12.500000  18.750000  25.000000  25.000000  18.750000     16
오리     0.000000  11.764706   0.000000  17.647059  17.647059  52.941176     17
독수리    0.000000  11.111111   0.000000   0.000000  33.333333  55.555556      9
타조     0.000000  10.000000   0.000000  20.000000  40.000000  30.000000     10
펭귄     0.000000   7.692308   0.000000  30.769231  30.769231  30.769231     13
말      0.000000   6.666667   6.666667   6.666667  26.666667  53.333333     15
개구리    0.000000   5.555556   5.555556   5.555556  44.444444  38.888889     18
햄스터    0.000000   0.000000  37.500000  12.500000  25.000000  25.000000      8
코알라    0.000000   0.000000  11.111111  33.333333  55.555556   0.000000      9
코끼리    0.000000   0.000000   9.090909  18.181818  27.272727  45.454545     11
호랑이    0.000000   0.000000   0.000000  28.571429  28.571429  42.857143      7
악어     0.000000   0.000000   0.000000  28.571429  14.285714  57.142857      7
곰      0.000000   0.000000   0.000000  20.000000  20.000000  60.000000     15
사자     0.000000   0.000000   0.000000  14.285714  14.285714  71.428571      7
원숭이    0.000000   0.000000   0.000000  12.500000  50.000000  37.500000      8
돼지     0.000000   0.000000   0.000000   7.142857   7.142857  85.714286     14
쥐      0.000000   0.000000   0.000000   6.666667  26.666667  66.666667     15
암소     0.000000   0.000000   0.000000   0.000000  75.000000  25.000000      4
닭      0.000000   0.000000   0.000000   0.000000  44.444444  55.555556      9
개미핥기   0.000000   0.000000   0.000000   0.000000  42.857143  57.142857      7
캥거루    0.000000   0.000000   0.000000   0.000000  37.500000  62.500000      8
수소     0.000000   0.000000   0.000000   0.000000  33.333333  66.666667      6
하마     0.000000   0.000000   0.000000   0.000000  28.571429  71.428571      7
새      0.000000   0.000000   0.000000   0.000000  23.076923  76.923077     13
고릴라    0.000000   0.000000   0.000000   0.000000  11.111111  88.888889      9

In [152]:
print(df_pct.sort_values(by=[1, 2, 3, 4, 5], ascending=False).to_string())

              1          2          3          4          5          6  count
문어    66.666667   0.000000  33.333333   0.000000   0.000000   0.000000      3
꼬마곰   18.750000   6.250000   6.250000  12.500000  31.250000  25.000000     16
코뿔소   16.666667   0.000000   0.000000  16.666667  50.000000  16.666667      6
염소    12.500000   0.000000  25.000000  25.000000   0.000000  37.500000      8
사슴    10.000000  30.000000  30.000000  30.000000   0.000000   0.000000     10
고양이    8.695652  26.086957  17.391304  17.391304  21.739130   8.695652     23
다람쥐    5.555556   5.555556  11.111111  22.222222  33.333333  22.222222     18
토끼     5.000000   0.000000  10.000000  15.000000  50.000000  20.000000     20
늑대     0.000000  18.181818  36.363636  27.272727  18.181818   0.000000     11
양      0.000000  15.384615   0.000000   7.692308  38.461538  38.461538     13
개      0.000000  12.500000  18.750000  25.000000  25.000000  18.750000     16
오리     0.000000  11.764706   0.000000  17.647059  17.647059  52.

다른 정렬 방식을 좀 연구해 봐야겠습니다.


In [153]:
df_pct['sort1'] = df_pct.loc[:, 1:2].sum(axis=1)
df_pct['sort2'] = df_pct.loc[:, 2:3].sum(axis=1)
df_pct['sort3'] = df_pct.loc[:, 3:4].sum(axis=1)
df_pct['sort4'] = df_pct.loc[:, 4:5].sum(axis=1)
df_pct['sort5'] = df_pct.loc[:, 5:6].sum(axis=1)
df_pct.sort_values(
  ['sort1', 'sort2', 'sort3', 'sort4', 'sort5'], ascending=False
).drop(columns=['sort1', 'sort2', 'sort3', 'sort4', 'sort5'])

1          2          3          4          5          6  count
문어    66.666667   0.000000  33.333333   0.000000   0.000000   0.000000      3
사슴    10.000000  30.000000  30.000000  30.000000   0.000000   0.000000     10
고양이    8.695652  26.086957  17.391304  17.391304  21.739130   8.695652     23
꼬마곰   18.750000   6.250000   6.250000  12.500000  31.250000  25.000000     16
늑대     0.000000  18.181818  36.363636  27.272727  18.181818   0.000000     11
코뿔소   16.666667   0.000000   0.000000  16.666667  50.000000  16.666667      6
양      0.000000  15.384615   0.000000   7.692308  38.461538  38.461538     13
개      0.000000  12.500000  18.750000  25.000000  25.000000  18.750000     16
염소    12.500000   0.000000  25.000000  25.000000   0.000000  37.500000      8
오리     0.000000  11.764706   0.000000  17.647059  17.647059  52.941176     17
다람쥐    5.555556   5.555556  11.111111  22.222222  33.333333  22.222222     18
독수리    0.000000  11.111111   0.000000   0.000000  33.333333  55.555556      9
타조     0.000000  10.000000   0.000000  20.000000  40.000000  30.000000     10
펭귄     0.000000   7.692308   0.000000  30.769231  30.769231  30.769231     13
말      0.000000   6.666667   6.666667   6.666667  26.666667  53.333333     15
개구리    0.000000   5.555556   5.555556   5.555556  44.444444  38.888889     18
토끼     5.000000   0.000000  10.000000  15.000000  50.000000  20.000000     20
햄스터    0.000000   0.000000  37.500000  12.500000  25.000000  25.000000      8
코알라    0.000000   0.000000  11.111111  33.333333  55.555556   0.000000      9
코끼리    0.000000   0.000000   9.090909  18.181818  27.272727  45.454545     11
호랑이    0.000000   0.000000   0.000000  28.571429  28.571429  42.857143      7
악어     0.000000   0.000000   0.000000  28.571429  14.285714  57.142857      7
곰      0.000000   0.000000   0.000000  20.000000  20.000000  60.000000     15
사자     0.000000   0.000000   0.000000  14.285714  14.285714  71.428571      7
원숭이    0.000000   0.000000   0.000000  12.500000  50.000000  37.500000      8
돼지     0.000000   0.000000   0.000000   7.142857   7.142857  85.714286     14
쥐      0.000000   0.000000   0.000000   6.666667  26.666667  66.666667     15
암소     0.000000   0.000000   0.000000   0.000000  75.000000  25.000000      4
닭      0.000000   0.000000   0.000000   0.000000  44.444444  55.555556      9
개미핥기   0.000000   0.000000   0.000000   0.000000  42.857143  57.142857      7
캥거루    0.000000   0.000000   0.000000   0.000000  37.500000  62.500000      8
수소     0.000000   0.000000   0.000000   0.000000  33.333333  66.666667      6
하마     0.000000   0.000000   0.000000   0.000000  28.571429  71.428571      7
새      0.000000   0.000000   0.000000   0.000000  23.076923  76.923077     13
고릴라    0.000000   0.000000   0.000000   0.000000  11.111111  88.888889      9

In [154]:
d = df_pct.sort_values(
  ['sort1', 'sort2', 'sort3', 'sort4', 'sort5'], ascending=False
).drop(
  columns=['sort1', 'sort2', 'sort3', 'sort4', 'sort5']
).to_dict(orient='records')

In [155]:
tiers = sorted(list(chars_t['tier'].unique()))
tiers

[1, 2, 3, 4, 5, 6]

In [156]:
species = list(df_pct.sort_values(
  ['sort1', 'sort2', 'sort3', 'sort4', 'sort5'], ascending=False
).index)
print(species)

['문어', '사슴', '고양이', '꼬마곰', '늑대', '코뿔소', '양', '개', '염소', '오리', '다람쥐', '독수리', '타조', '펭귄', '말', '개구리', '토끼', '햄스터', '코알라', '코끼리', '호랑이', '악어', '곰', '사자', '원숭이', '돼지', '쥐', '암소', '닭', '개미핥기', '캥거루', '수소', '하마', '새', '고릴라']


In [157]:
[[e[1], e[2], e[3], e[4], e[5], e[6]] for e in d]

[[66.66666666666666, 0.0, 33.33333333333333, 0.0, 0.0, 0.0],
 [10.0, 30.0, 30.0, 30.0, 0.0, 0.0],
 [8.695652173913043,
  26.08695652173913,
  17.391304347826086,
  17.391304347826086,
  21.73913043478261,
  8.695652173913043],
 [18.75, 6.25, 6.25, 12.5, 31.25, 25.0],
 [0.0,
  18.181818181818183,
  36.36363636363637,
  27.27272727272727,
  18.181818181818183,
  0.0],
 [16.666666666666664, 0.0, 0.0, 16.666666666666664, 50.0, 16.666666666666664],
 [0.0,
  15.384615384615385,
  0.0,
  7.6923076923076925,
  38.46153846153847,
  38.46153846153847],
 [0.0, 12.5, 18.75, 25.0, 25.0, 18.75],
 [12.5, 0.0, 25.0, 25.0, 0.0, 37.5],
 [0.0,
  11.76470588235294,
  0.0,
  17.647058823529413,
  17.647058823529413,
  52.94117647058824],
 [5.555555555555555,
  5.555555555555555,
  11.11111111111111,
  22.22222222222222,
  33.33333333333333,
  22.22222222222222],
 [0.0, 11.11111111111111, 0.0, 0.0, 33.33333333333333, 55.55555555555556],
 [0.0, 10.0, 0.0, 20.0, 40.0, 30.0],
 [0.0,
  7.6923076923076925,
  0.0

In [158]:
import math


In [159]:
round(1.5362351235, 2)

1.54

In [160]:
[[round(e[1], 2), round(e[2], 2), round(e[3], 2), round(e[4], 2), round(e[5], 2), round(e[6], 2)] for e in d]

[[66.67, 0.0, 33.33, 0.0, 0.0, 0.0],
 [10.0, 30.0, 30.0, 30.0, 0.0, 0.0],
 [8.7, 26.09, 17.39, 17.39, 21.74, 8.7],
 [18.75, 6.25, 6.25, 12.5, 31.25, 25.0],
 [0.0, 18.18, 36.36, 27.27, 18.18, 0.0],
 [16.67, 0.0, 0.0, 16.67, 50.0, 16.67],
 [0.0, 15.38, 0.0, 7.69, 38.46, 38.46],
 [0.0, 12.5, 18.75, 25.0, 25.0, 18.75],
 [12.5, 0.0, 25.0, 25.0, 0.0, 37.5],
 [0.0, 11.76, 0.0, 17.65, 17.65, 52.94],
 [5.56, 5.56, 11.11, 22.22, 33.33, 22.22],
 [0.0, 11.11, 0.0, 0.0, 33.33, 55.56],
 [0.0, 10.0, 0.0, 20.0, 40.0, 30.0],
 [0.0, 7.69, 0.0, 30.77, 30.77, 30.77],
 [0.0, 6.67, 6.67, 6.67, 26.67, 53.33],
 [0.0, 5.56, 5.56, 5.56, 44.44, 38.89],
 [5.0, 0.0, 10.0, 15.0, 50.0, 20.0],
 [0.0, 0.0, 37.5, 12.5, 25.0, 25.0],
 [0.0, 0.0, 11.11, 33.33, 55.56, 0.0],
 [0.0, 0.0, 9.09, 18.18, 27.27, 45.45],
 [0.0, 0.0, 0.0, 28.57, 28.57, 42.86],
 [0.0, 0.0, 0.0, 28.57, 14.29, 57.14],
 [0.0, 0.0, 0.0, 20.0, 20.0, 60.0],
 [0.0, 0.0, 0.0, 14.29, 14.29, 71.43],
 [0.0, 0.0, 0.0, 12.5, 50.0, 37.5],
 [0.0, 0.0, 0.0, 7.14, 7

안지우님이 가로세로를 바꿔달라고 하셨습니다.


In [161]:
from operator import itemgetter

In [164]:
[[round(v[i], 2) for v in d] for i in range(1, 7)]


[[66.67,
  10.0,
  8.7,
  18.75,
  0.0,
  16.67,
  0.0,
  0.0,
  12.5,
  0.0,
  5.56,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  5.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0],
 [0.0,
  30.0,
  26.09,
  6.25,
  18.18,
  0.0,
  15.38,
  12.5,
  0.0,
  11.76,
  5.56,
  11.11,
  10.0,
  7.69,
  6.67,
  5.56,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0],
 [33.33,
  30.0,
  17.39,
  6.25,
  36.36,
  0.0,
  0.0,
  18.75,
  25.0,
  0.0,
  11.11,
  0.0,
  0.0,
  0.0,
  6.67,
  5.56,
  10.0,
  37.5,
  11.11,
  9.09,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0],
 [0.0,
  30.0,
  17.39,
  12.5,
  27.27,
  16.67,
  7.69,
  25.0,
  25.0,
  17.65,
  22.22,
  0.0,
  20.0,
  30.77,
  6.67,
  5.56,
  15.0,
  12.5,
  33.33,
  18.18,
  28.57,
  28.57,
  20.0,
  14.29,
  12.5,
  7.14

In [163]:
d


[{1: 66.66666666666666,
  2: 0.0,
  3: 33.33333333333333,
  4: 0.0,
  5: 0.0,
  6: 0.0,
  'count': 3},
 {1: 10.0, 2: 30.0, 3: 30.0, 4: 30.0, 5: 0.0, 6: 0.0, 'count': 10},
 {1: 8.695652173913043,
  2: 26.08695652173913,
  3: 17.391304347826086,
  4: 17.391304347826086,
  5: 21.73913043478261,
  6: 8.695652173913043,
  'count': 23},
 {1: 18.75, 2: 6.25, 3: 6.25, 4: 12.5, 5: 31.25, 6: 25.0, 'count': 16},
 {1: 0.0,
  2: 18.181818181818183,
  3: 36.36363636363637,
  4: 27.27272727272727,
  5: 18.181818181818183,
  6: 0.0,
  'count': 11},
 {1: 16.666666666666664,
  2: 0.0,
  3: 0.0,
  4: 16.666666666666664,
  5: 50.0,
  6: 16.666666666666664,
  'count': 6},
 {1: 0.0,
  2: 15.384615384615385,
  3: 0.0,
  4: 7.6923076923076925,
  5: 38.46153846153847,
  6: 38.46153846153847,
  'count': 13},
 {1: 0.0, 2: 12.5, 3: 18.75, 4: 25.0, 5: 25.0, 6: 18.75, 'count': 16},
 {1: 12.5, 2: 0.0, 3: 25.0, 4: 25.0, 5: 0.0, 6: 37.5, 'count': 8},
 {1: 0.0,
  2: 11.76470588235294,
  3: 0.0,
  4: 17.647058823529413,